xla/service/gpu/openxla/compiler.cc

```diff
@@ -0,0 +1,235 @@
+
+OpenXlaCompiler::Bytecode::Bytecode(iree_compiler_output_t* output, void* data,
+                                    size_t length)
+    : output_(output), data_(data), length_(length) {}
+
+OpenXlaCompiler::Bytecode::~Bytecode() { ireeCompilerOutputDestroy(output_); }
+
+//===-----------------------------------------------------------------------===/
+// OpenXlaCompiler
+//===-----------------------------------------------------------------------===/
+
+OpenXlaCompiler::OpenXlaCompiler(iree_compiler_session_t* session,
+                                 iree_compiler_invocation_t* inv)
+    : session_(session), inv_(inv) {}
+
+OpenXlaCompiler::~OpenXlaCompiler() {
+  if (error_) {
+    ireeCompilerErrorDestroy(error_);
+  }
+
+  ireeCompilerInvocationDestroy(inv_);
+  ireeCompilerSessionDestroy(session_);
+
+  if (output_) {
+    ireeCompilerOutputDestroy(output_);
+  }
+}
+
+bool OpenXlaCompiler::ParseSourceBuffer(std::string_view buffer) {
+  iree_compiler_source_t* source;
+  auto* error = ireeCompilerSourceWrapBuffer(
+      session_, "<jit>", buffer.data(), buffer.size(),
+      /*isNullTerminated=*/false, &source);
+  if (error) {
+    SetError(error);
+    return false;
+  }
+
+  return ireeCompilerInvocationParseSource(inv_, source);
+}
+
+bool OpenXlaCompiler::SetFlag(const char* flag) {
+  auto* error = ireeCompilerSessionSetFlags(session_, 1, &flag);
+  if (error) {
+    SetError(error);
+    return false;
+  }
+  return true;
+}
+
+std::unique_ptr<OpenXlaCompiler::Bytecode>
+OpenXlaCompiler::CompileStandardPipeline() {
+  if (!ireeCompilerInvocationPipeline(inv_, IREE_COMPILER_PIPELINE_STD)) {
+    return nullptr;
+  }
+
+  iree_compiler_error_t* error = ireeCompilerOutputOpenMembuffer(&output_);
+  if (error) {
+    SetError(error);
+    return nullptr;
+  }
+
+  error = ireeCompilerInvocationOutputVMBytecode(inv_, output_);
+  if (error) {
+    SetError(error);
+    return nullptr;
+  }
+
+  void* output_data = nullptr;
+  uint64_t size;
+  error = ireeCompilerOutputMapMemory(output_, &output_data, &size);
+  if (error) {
+    SetError(error);
+    return nullptr;
+  }
+
+  // Transfer the output_ to Bytecode since the mapping is only
+  // valid for the life of the output.
+  iree_compiler_output_t* local_output = output_;
+  output_ = nullptr;
+  return std::make_unique<Bytecode>(local_output, output_data, size);
+}
+
+//===-----------------------------------------------------------------------===/
+// Loading OpenXlaCompiler from a library
+//===-----------------------------------------------------------------------===/
+
+static bool InitializeCompilerForProcess(const std::string& library_path) {
+  if (!ireeCompilerLoadLibrary(library_path.c_str())) {
+    return false;
+  }
+
+  ireeCompilerGlobalInitialize();
+  return true;
+}
+
+static std::optional<std::string_view> LoadCompilerStubOnce(
+    const std::string& library_path) {
+  static std::string* loaded_path = [&]() -> std::string* {
```

Comment:
Could also use [`call_once`](https://en.cppreference.com/w/cpp/thread/call_once) (or some similar utility) rather than `static`, if you want to allow for the rest of the application to load before this code runs.

Reply:
Replaced with `call_once` as it makes intention more explicit, but static ptr + lambda actually run only on first call, it's not the same as static initializer running at load time

Reply:
Ah good to know (so many years writing C++ code and initialization still confounds me :P)